## Final Project Submission

Please fill out:
* Student name: Stanoff Kipkirui
* Student pace:  full time
* Scheduled project review date/time: 03/07/2023 to 09/07/2023
* Instructor name:Veronica Isiaho 
* Blog post URL:


    
    **USE OF MULTIPLE LINEAR REGRESSION MODELING TO ANALYZE HOUSE SALES IN KING COUNTY**
                        

**INTRODUCTION**

In this data science project, we aim to build a multiple linear regression model to predict house sale prices in King County, leveraging the King County House Sales dataset. The project's objective is to provide accurate estimates of property values to homeowners and potential buyers, enabling informed decision-making in the real estate market.

**BUSINESS UNDERSTANDING**

The real estate agency needs to understand the relationship between various features of the houses and their sale prices in order to provide accurate advice to homeowners. By performing multiple linear regression analysis on the King County House Sales dataset, the agency can identify which home renovation factors significantly impact the sale price.

**DATA UNDERSTANDING**

The King County House Sales dataset contains comprehensive information about house sales, including features such as the number of bedrooms, number of bathrooms, square footage of living space in the home, number of floors in the house, how good the overall condition of the house is, overall house grade which is related to design and construction, and year when the house was built. To better understand the dataset, we will perform exploratory data analysis, examine distributions, identify correlations, and assess feature importance.

**DATA GATHERING**

Since we are using python programming language for data science we need to import python packages that will be utilized in this study and then load the datasets relevant for our study.

In [1]:
#python packages that will be utilized in this project
import statsmodels.api as sm
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np
import seaborn as sns
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import StandardScaler,PolynomialFeatures
from sklearn.linear_model import LinearRegression
%matplotlib inline

https://www.kaggle.com/code/vishnuramachandran/data-analysis-python-house-sales-in-king-county

In [2]:
#load King County House Sale dataset

data = pd.read_csv("data/kc_house_data.csv")
data.head()

id        date     price  bedrooms  bathrooms  sqft_living  \
0  7129300520  10/13/2014  221900.0         3       1.00         1180   
1  6414100192   12/9/2014  538000.0         3       2.25         2570   
2  5631500400   2/25/2015  180000.0         2       1.00          770   
3  2487200875   12/9/2014  604000.0         4       3.00         1960   
4  1954400510   2/18/2015  510000.0         3       2.00         1680   

   sqft_lot  floors waterfront  view  ...          grade sqft_above  \
0      5650     1.0        NaN  NONE  ...      7 Average       1180   
1      7242     2.0         NO  NONE  ...      7 Average       2170   
2     10000     1.0         NO  NONE  ...  6 Low Average        770   
3      5000     1.0         NO  NONE  ...      7 Average       1050   
4      8080     1.0         NO  NONE  ...         8 Good       1680   

   sqft_basement yr_built  yr_renovated  zipcode      lat     long  \
0            0.0     1955           0.0    98178  47.5112 -122.257   
1          400.0     1951        1991.0    98125  47.7210 -122.319   
2            0.0     1933           NaN    98028  47.7379 -122.233   
3          910.0     1965           0.0    98136  47.5208 -122.393   
4            0.0     1987           0.0    98074  47.6168 -122.045   

   sqft_living15  sqft_lot15  
0           1340        5650  
1           1690        7639  
2           2720        8062  
3           1360        5000  
4           1800        7503  

[5 rows x 21 columns]

In [3]:
#get summary statistics for our dataset
data.describe()

id         price      bedrooms     bathrooms   sqft_living  \
count  2.159700e+04  2.159700e+04  21597.000000  21597.000000  21597.000000   
mean   4.580474e+09  5.402966e+05      3.373200      2.115826   2080.321850   
std    2.876736e+09  3.673681e+05      0.926299      0.768984    918.106125   
min    1.000102e+06  7.800000e+04      1.000000      0.500000    370.000000   
25%    2.123049e+09  3.220000e+05      3.000000      1.750000   1430.000000   
50%    3.904930e+09  4.500000e+05      3.000000      2.250000   1910.000000   
75%    7.308900e+09  6.450000e+05      4.000000      2.500000   2550.000000   
max    9.900000e+09  7.700000e+06     33.000000      8.000000  13540.000000   

           sqft_lot        floors    sqft_above      yr_built  yr_renovated  \
count  2.159700e+04  21597.000000  21597.000000  21597.000000  17755.000000   
mean   1.509941e+04      1.494096   1788.596842   1970.999676     83.636778   
std    4.141264e+04      0.539683    827.759761     29.375234    399.946414   
min    5.200000e+02      1.000000    370.000000   1900.000000      0.000000   
25%    5.040000e+03      1.000000   1190.000000   1951.000000      0.000000   
50%    7.618000e+03      1.500000   1560.000000   1975.000000      0.000000   
75%    1.068500e+04      2.000000   2210.000000   1997.000000      0.000000   
max    1.651359e+06      3.500000   9410.000000   2015.000000   2015.000000   

            zipcode           lat          long  sqft_living15     sqft_lot15  
count  21597.000000  21597.000000  21597.000000   21597.000000   21597.000000  
mean   98077.951845     47.560093   -122.213982    1986.620318   12758.283512  
std       53.513072      0.138552      0.140724     685.230472   27274.441950  
min    98001.000000     47.155900   -122.519000     399.000000     651.000000  
25%    98033.000000     47.471100   -122.328000    1490.000000    5100.000000  
50%    98065.000000     47.571800   -122.231000    1840.000000    7620.000000  
75%    98118.000000     47.678000   -122.125000    2360.000000   10083.000000  
max    98199.000000     47.777600   -121.315000    6210.000000  871200.000000

**DATA CLEANING AND PREPARATION**

This involves checking data to check if there exists any possible problems that makes data unsuitable for analysis.The problems may include missing data , duplicated data , handling categorical variables and obtaining the relevant subset from our dataset  that are likely to have a significant impact on the target variable.Data cleaning and preparation done in our datasets include checking for missing values, dropping unnecessary columns, getting rid of duplicates,handling categorical variables and creating new dataframes that suit our study including subsetting our datasets.Each datacleaning process is described in details in each cell based on the dataset.
Feature Selection:
    
Select relevant features that are likely to have a significant impact on the target variable (sale price). Consider both numerical and categorical variables based on domain knowledge and correlation analysis. Drop any unnecessary or redundant columns.I will focus on a subset of the overall dataset.These features are:

price:sale price which is our prediction target
bedrooms:number of bed rooms in the house
bathrooms:number of bathrooms in the house
sqft_living:square footage of living space in the home
sqft_lot:square footage of the lot
floors:number of floors or the levels in the house
waterfront:whether the house is located on a water front
grade:overall house grade in relation to design and construction


In [4]:
data_subset =  data[['price', 'bedrooms', 'bathrooms', 'sqft_living', 'sqft_lot', 'floors', 'waterfront', 'grade']].copy()
data_subset

price  bedrooms  bathrooms  sqft_living  sqft_lot  floors  \
0      221900.0         3       1.00         1180      5650     1.0   
1      538000.0         3       2.25         2570      7242     2.0   
2      180000.0         2       1.00          770     10000     1.0   
3      604000.0         4       3.00         1960      5000     1.0   
4      510000.0         3       2.00         1680      8080     1.0   
...         ...       ...        ...          ...       ...     ...   
21592  360000.0         3       2.50         1530      1131     3.0   
21593  400000.0         4       2.50         2310      5813     2.0   
21594  402101.0         2       0.75         1020      1350     2.0   
21595  400000.0         3       2.50         1600      2388     2.0   
21596  325000.0         2       0.75         1020      1076     2.0   

      waterfront          grade  
0            NaN      7 Average  
1             NO      7 Average  
2             NO  6 Low Average  
3             NO      7 Average  
4             NO         8 Good  
...          ...            ...  
21592         NO         8 Good  
21593         NO         8 Good  
21594         NO      7 Average  
21595        NaN         8 Good  
21596         NO      7 Average  

[21597 rows x 8 columns]

In [5]:
#get summary statistics for our dataset
data_subset.describe()

price      bedrooms     bathrooms   sqft_living      sqft_lot  \
count  2.159700e+04  21597.000000  21597.000000  21597.000000  2.159700e+04   
mean   5.402966e+05      3.373200      2.115826   2080.321850  1.509941e+04   
std    3.673681e+05      0.926299      0.768984    918.106125  4.141264e+04   
min    7.800000e+04      1.000000      0.500000    370.000000  5.200000e+02   
25%    3.220000e+05      3.000000      1.750000   1430.000000  5.040000e+03   
50%    4.500000e+05      3.000000      2.250000   1910.000000  7.618000e+03   
75%    6.450000e+05      4.000000      2.500000   2550.000000  1.068500e+04   
max    7.700000e+06     33.000000      8.000000  13540.000000  1.651359e+06   

             floors  
count  21597.000000  
mean       1.494096  
std        0.539683  
min        1.000000  
25%        1.000000  
50%        1.500000  
75%        2.000000  
max        3.500000

In handling missing values we can observe from the summary statistics that the count for each column is 21597 which gives an implication that we do not have any missing data. So we are certain that we have evaded the dangers of performing multiple linear regression with missing data.Handling missing values before performing multiple linear regression is essential to ensure unbiased and accurate results, preserve sample size and statistical power, maintain data integrity, fulfill the assumptions of linear regression, enhance model performance, and enable fair comparisons and generalizability of the findings.



Checking for duplicates:
    
Duplicate observations can affect the regression model's results, as they may introduce bias and impact the statistical properties of the model.So we first check for duplicates and drop the duplicates but keep the first occurence.By dropping the duplicates, we ensure that each observation in our dataset is unique, which is important for reliable and accurate analysis in multiple linear regression.

In [6]:
#  First heck for duplicates
duplicate_rows = data_subset.duplicated()
num_duplicates = duplicate_rows.sum()
num_duplicates
# Display the duplicate rows 
duplicate_data = data_subset[duplicate_rows]
#print("Duplicate rows:")
print(duplicate_data)
# Drop duplicates and keep the first occurrence
data_subset.drop_duplicates(keep='first', inplace=True)

# Verify if duplicates have been dropped
num_duplicates = data_subset.duplicated().sum()
num_duplicates

          price  bedrooms  bathrooms  sqft_living  sqft_lot  floors  \
4348   259950.0         2       2.00         1070       649     2.0   
14969  585000.0         3       2.50         2290      5089     2.0   
17228  629950.0         3       2.50         1680      1683     2.0   
20038  555000.0         3       2.50         1940      3211     2.0   
20508  529500.0         3       2.25         1410       905     3.0   
21458  359800.0         5       2.50         2170      2752     2.0   
21518  599000.0         3       1.75         1650      1180     3.0   

      waterfront     grade  
4348          NO  9 Better  
14969         NO  9 Better  
17228         NO  9 Better  
20038         NO    8 Good  
20508         NO  9 Better  
21458         NO    8 Good  
21518         NO    8 Good  


0

Handling Categorical Variables:
First we inspect pandas data types in our dataset to know whethervwe will use all te columns.Object data type cannot be used because the model will crash.Convert categorical variables into numerical representations using technique called one-hot encoding. This ensures that the data can be properly utilized in the regression model.

In [7]:
data_subset.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 21590 entries, 0 to 21596
Data columns (total 8 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   price        21590 non-null  float64
 1   bedrooms     21590 non-null  int64  
 2   bathrooms    21590 non-null  float64
 3   sqft_living  21590 non-null  int64  
 4   sqft_lot     21590 non-null  int64  
 5   floors       21590 non-null  float64
 6   waterfront   19214 non-null  object 
 7   grade        21590 non-null  object 
dtypes: float64(3), int64(3), object(2)
memory usage: 1.5+ MB


In [8]:
# One-hot encoding for categorical variables
data_encoded = pd.get_dummies(data_subset, columns=['waterfront','grade'])
data_encoded


price  bedrooms  bathrooms  sqft_living  sqft_lot  floors  \
0      221900.0         3       1.00         1180      5650     1.0   
1      538000.0         3       2.25         2570      7242     2.0   
2      180000.0         2       1.00          770     10000     1.0   
3      604000.0         4       3.00         1960      5000     1.0   
4      510000.0         3       2.00         1680      8080     1.0   
...         ...       ...        ...          ...       ...     ...   
21592  360000.0         3       2.50         1530      1131     3.0   
21593  400000.0         4       2.50         2310      5813     2.0   
21594  402101.0         2       0.75         1020      1350     2.0   
21595  400000.0         3       2.50         1600      2388     2.0   
21596  325000.0         2       0.75         1020      1076     2.0   

       waterfront_NO  waterfront_YES  grade_10 Very Good  grade_11 Excellent  \
0                  0               0                   0                   0   
1                  1               0                   0                   0   
2                  1               0                   0                   0   
3                  1               0                   0                   0   
4                  1               0                   0                   0   
...              ...             ...                 ...                 ...   
21592              1               0                   0                   0   
21593              1               0                   0                   0   
21594              1               0                   0                   0   
21595              0               0                   0                   0   
21596              1               0                   0                   0   

       grade_12 Luxury  grade_13 Mansion  grade_3 Poor  grade_4 Low  \
0                    0                 0             0            0   
1                    0                 0             0            0   
2                    0                 0             0            0   
3                    0                 0             0            0   
4                    0                 0             0            0   
...                ...               ...           ...          ...   
21592                0                 0             0            0   
21593                0                 0             0            0   
21594                0                 0             0            0   
21595                0                 0             0            0   
21596                0                 0             0            0   

       grade_5 Fair  grade_6 Low Average  grade_7 Average  grade_8 Good  \
0                 0                    0                1             0   
1                 0                    0                1             0   
2                 0                    1                0             0   
3                 0                    0                1             0   
4                 0                    0                0             1   
...             ...                  ...              ...           ...   
21592             0                    0                0             1   
21593             0                    0                0             1   
21594             0                    0                1             0   
21595             0                    0                0             1   
21596             0                    0                1             0   

       grade_9 Better  
0                   0  
1                   0  
2                   0  
3                   0  
4                   0  
...               ...  
21592               0  
21593               0  
21594               0  
21595               0  
21596               0  

[21590 rows x 19 columns]

In [ ]:
# Standardization using z-score normalization
#from sklearn.preprocessing import StandardScaler

#scaler = StandardScaler()
#data_scaled = scaler.fit_transform(data_encoded)
#data_scaled = pd.DataFrame(data_scaled, columns=data_encoded.columns)


Our data is now clean and ready for analysis

In [ ]:
**MODELING AND DATA ANALYSIS**

